sphinxcontrib/confluencebuilder/translator/wiki.py

```diff
@@ -870,10 +866,18 @@ def visit_reference(self, node):
             label = node.astext()
             for find, encoded in SPECIAL_VALUE_REPLACEMENTS:
                 label = label.replace(find, encoded)
-            if label == doctitle and not anchor:
+            if not doctitle:
```

Comment:
While I don't mind mixing related changes in a single PR (in this case, adding more in-depth support for self-page anchors and todo support), I would extremely recommend splitting this up into multiple commits so it's easier to reflect back on changes made when looking at older commit messages.

It appears to me you've removed the top `doctitle` check to help generate anchors on the same document for documents which don't have a title value (side note, title-less documents will never be included in part of the publish event). I believe this introduces two possible issues. First, when `refuri` is set, it should be (assumption) that the reference is to another document (maybe itself) in the doctree. In the event that a document references another document without a title, instead of generating a warning message an invalid anchor reference will be created (e.x. doca to docb#section could generate an anchor link in doca called `[#section]` if docb has no title). Second issue is that in the event a page links to another title-less page without an anchor, `anchor` will be empty and the `assert` will be triggered.

If you really want to permit support anchor references contained within a single document with no title, there should be a check if `docname == self.docname` before making the assumption that an internal anchor will work (maybe only also permitting this case when an anchor value is detected as well).

Reply:
I'll see if I can come up with some tests and examples that hit this.  Nothing seemed to hit this code and as per my comment on this PR, I was surprised to find that it was required.  Bit busy with GCSEs and A-levels at the moment (my daughters not mine :-) ) so apologies for not having looked at this sooner.